File: session_py/src/session_py/pointcloud_test.py
from .pointcloud import PointCloud
from .point import Point
from .vector import Vector
from .color import Color


def test_pointcloud_new():
    points = [
        Point(0.0, 0.0, 0.0),
        Point(1.0, 0.0, 0.0),
        Point(0.0, 1.0, 0.0),
    ]
    normals = [
        Vector(0.0, 0.0, 1.0),
        Vector(0.0, 1.0, 0.0),
        Vector(1.0, 0.0, 0.0),
    ]
    colors = [
        Color(255, 0, 0, 255),
        Color(0, 255, 0, 255),
        Color(0, 0, 255, 255),
    ]
    cloud = PointCloud(points, normals, colors)
    assert len(cloud) == 3
    assert not cloud.is_empty()


def test_pointcloud_default():
    cloud = PointCloud()
    assert len(cloud) == 0
    assert cloud.is_empty()
    assert cloud.name == "my_pointcloud"


def test_pointcloud_iadd_vector():
    cloud = PointCloud(
        [Point(1.0, 2.0, 3.0)],
        [Vector(0.0, 0.0, 1.0)],
        [Color(255, 0, 0, 255)],
    )
    v = Vector(4.0, 5.0, 6.0)
    cloud += v
    assert cloud.points[0].x == 5.0
    assert cloud.points[0].y == 7.0
    assert cloud.points[0].z == 9.0


def test_pointcloud_add_vector():
    cloud = PointCloud(
        [Point(1.0, 2.0, 3.0)],
        [Vector(0.0, 0.0, 1.0)],
        [Color(255, 0, 0, 255)],
    )
    v = Vector(4.0, 5.0, 6.0)
    cloud2 = cloud + v
    assert cloud2.points[0].x == 5.0
    assert cloud2.points[0].y == 7.0
    assert cloud2.points[0].z == 9.0


def test_pointcloud_isub_vector():
    cloud = PointCloud(
        [Point(1.0, 2.0, 3.0)],
        [Vector(0.0, 0.0, 1.0)],
        [Color(255, 0, 0, 255)],
    )
    v = Vector(4.0, 5.0, 6.0)
    cloud -= v
    assert cloud.points[0].x == -3.0
    assert cloud.points[0].y == -3.0
    assert cloud.points[0].z == -3.0


def test_pointcloud_sub_vector():
    cloud = PointCloud(
        [Point(1.0, 2.0, 3.0)],
        [Vector(0.0, 0.0, 1.0)],
        [Color(255, 0, 0, 255)],
    )
    v = Vector(4.0, 5.0, 6.0)
    cloud2 = cloud - v
    assert cloud2.points[0].x == -3.0
    assert cloud2.points[0].y == -3.0
    assert cloud2.points[0].z == -3.0


def test_pointcloud_str():
    cloud = PointCloud(
        [Point(0.0, 0.0, 0.0)],
        [Vector(0.0, 0.0, 1.0)],
        [Color(255, 0, 0, 255)],
    )
    s = str(cloud)
    assert "PointCloud" in s
    assert "points=1" in s


def test_pointcloud_json_serialization():
    cloud = PointCloud(
        [Point(1.0, 2.0, 3.0)],
        [Vector(0.0, 0.0, 1.0)],
        [Color(255, 0, 0, 255)],
    )
    json_data = cloud.to_json_data()
    cloud2 = PointCloud.from_json_data(json_data)
    assert cloud2.points[0].x == 1.0
    assert cloud2.points[0].y == 2.0
    assert cloud2.points[0].z == 3.0


def test_pointcloud_json_file():
    cloud = PointCloud(
        [
            Point(1.0, 2.0, 3.0),
            Point(4.0, 5.0, 6.0),
            Point(7.0, 8.0, 9.0),
        ],
        [
            Vector(0.0, 0.0, 1.0),
            Vector(0.0, 1.0, 0.0),
            Vector(1.0, 0.0, 0.0),
        ],
        [
            Color(255, 0, 0, 255),
            Color(0, 255, 0, 255),
            Color(0, 0, 255, 255),
        ],
    )
    cloud.to_json("test_pointcloud.json")
    cloud2 = PointCloud.from_json("test_pointcloud.json")

    assert len(cloud2) == 3
    assert cloud2.points[0].x == 1.0
    assert cloud2.points[1].y == 5.0
    assert cloud2.points[2].z == 9.0


def test_pointcloud_json_multiple_points():
    cloud = PointCloud(
        [
            Point(1.0, 2.0, 3.0),
            Point(4.0, 5.0, 6.0),
            Point(7.0, 8.0, 9.0),
        ],
        [
            Vector(0.0, 0.0, 1.0),
            Vector(0.0, 1.0, 0.0),
            Vector(1.0, 0.0, 0.0),
        ],
        [
            Color(255, 0, 0, 255),
            Color(0, 255, 0, 255),
            Color(0, 0, 255, 255),
        ],
    )
    json_data = cloud.to_json_data()
    cloud2 = PointCloud.from_json_data(json_data)

    assert len(cloud2) == 3
    assert cloud2.points[0].x == 1.0
    assert cloud2.points[1].y == 5.0
    assert cloud2.points[2].z == 9.0
    assert cloud2.normals[0].z == 1.0
    assert cloud2.colors[1].g == 255
    # Verify alpha is always 255 after deserialization
    assert cloud2.colors[0].a == 255
    assert cloud2.colors[1].a == 255
    assert cloud2.colors[2].a == 255

File: session_cpp/src/pointcloud_test.cpp
#include "pointcloud.h"
#include "catch_amalgamated.hpp"

using namespace session_cpp;

TEST_CASE("PointCloud new") {
    std::vector<Point> points = {
        Point(0.0f, 0.0f, 0.0f),
        Point(1.0f, 0.0f, 0.0f),
        Point(0.0f, 1.0f, 0.0f)
    };
    std::vector<Vector> normals = {
        Vector(0.0f, 0.0f, 1.0f),
        Vector(0.0f, 1.0f, 0.0f),
        Vector(1.0f, 0.0f, 0.0f)
    };
    std::vector<Color> colors = {
        Color(255, 0, 0, 255),
        Color(0, 255, 0, 255),
        Color(0, 0, 255, 255)
    };
    PointCloud cloud(points, normals, colors);
    REQUIRE(cloud.size() == 3);
    REQUIRE(!cloud.empty());
}

TEST_CASE("PointCloud default") {
    PointCloud cloud;
    REQUIRE(cloud.size() == 0);
    REQUIRE(cloud.empty());
    REQUIRE(cloud.name() == "my_pointcloud");
}

TEST_CASE("PointCloud operator+= vector") {
    std::vector<Point> points = {Point(1.0f, 2.0f, 3.0f)};
    std::vector<Vector> normals = {Vector(0.0f, 0.0f, 1.0f)};
    std::vector<Color> colors = {Color(255, 0, 0, 255)};
    PointCloud cloud(points, normals, colors);
    
    Vector v(4.0f, 5.0f, 6.0f);
    cloud += v;
    
    REQUIRE(cloud.points()[0].x() == 5.0f);
    REQUIRE(cloud.points()[0].y() == 7.0f);
    REQUIRE(cloud.points()[0].z() == 9.0f);
}

TEST_CASE("PointCloud operator+ vector") {
    std::vector<Point> points = {Point(1.0f, 2.0f, 3.0f)};
    std::vector<Vector> normals = {Vector(0.0f, 0.0f, 1.0f)};
    std::vector<Color> colors = {Color(255, 0, 0, 255)};
    PointCloud cloud(points, normals, colors);
    
    Vector v(4.0f, 5.0f, 6.0f);
    PointCloud cloud2 = cloud + v;
    
    REQUIRE(cloud2.points()[0].x() == 5.0f);
    REQUIRE(cloud2.points()[0].y() == 7.0f);
    REQUIRE(cloud2.points()[0].z() == 9.0f);
}

TEST_CASE("PointCloud operator-= vector") {
    std::vector<Point> points = {Point(1.0f, 2.0f, 3.0f)};
    std::vector<Vector> normals = {Vector(0.0f, 0.0f, 1.0f)};
    std::vector<Color> colors = {Color(255, 0, 0, 255)};
    PointCloud cloud(points, normals, colors);
    
    Vector v(4.0f, 5.0f, 6.0f);
    cloud -= v;
    
    REQUIRE(cloud.points()[0].x() == -3.0f);
    REQUIRE(cloud.points()[0].y() == -3.0f);
    REQUIRE(cloud.points()[0].z() == -3.0f);
}

TEST_CASE("PointCloud operator- vector") {
    std::vector<Point> points = {Point(1.0f, 2.0f, 3.0f)};
    std::vector<Vector> normals = {Vector(0.0f, 0.0f, 1.0f)};
    std::vector<Color> colors = {Color(255, 0, 0, 255)};
    PointCloud cloud(points, normals, colors);
    
    Vector v(4.0f, 5.0f, 6.0f);
    PointCloud cloud2 = cloud - v;
    
    REQUIRE(cloud2.points()[0].x() == -3.0f);
    REQUIRE(cloud2.points()[0].y() == -3.0f);
    REQUIRE(cloud2.points()[0].z() == -3.0f);
}

TEST_CASE("PointCloud to_string") {
    std::vector<Point> points = {Point(0.0f, 0.0f, 0.0f)};
    std::vector<Vector> normals = {Vector(0.0f, 0.0f, 1.0f)};
    std::vector<Color> colors = {Color(255, 0, 0, 255)};
    PointCloud cloud(points, normals, colors);
    
    std::string str = cloud.to_string();
    REQUIRE(str.find("PointCloud") != std::string::npos);
    REQUIRE(str.find("points=1") != std::string::npos);
}

TEST_CASE("PointCloud JSON serialization") {
    std::vector<Point> points = {Point(1.0f, 2.0f, 3.0f)};
    std::vector<Vector> normals = {Vector(0.0f, 0.0f, 1.0f)};
    std::vector<Color> colors = {Color(255, 0, 0, 255)};
    PointCloud cloud(points, normals, colors);
    
    auto json = cloud.to_json_data();
    PointCloud cloud2 = PointCloud::from_json_data(json);
    
    REQUIRE(cloud2.points()[0].x() == 1.0f);
    REQUIRE(cloud2.points()[0].y() == 2.0f);
    REQUIRE(cloud2.points()[0].z() == 3.0f);
}

TEST_CASE("PointCloud JSON file") {
    std::vector<Point> points = {
        Point(1.0f, 2.0f, 3.0f),
        Point(4.0f, 5.0f, 6.0f),
        Point(7.0f, 8.0f, 9.0f)
    };
    std::vector<Vector> normals = {
        Vector(0.0f, 0.0f, 1.0f),
        Vector(0.0f, 1.0f, 0.0f),
        Vector(1.0f, 0.0f, 0.0f)
    };
    std::vector<Color> colors = {
        Color(255, 0, 0, 255),
        Color(0, 255, 0, 255),
        Color(0, 0, 255, 255)
    };
    PointCloud cloud(points, normals, colors);
    
    cloud.to_json("../test_pointcloud.json");
    PointCloud cloud2 = PointCloud::from_json("../test_pointcloud.json");
    
    REQUIRE(cloud2.size() == 3);
    REQUIRE(cloud2.points()[0].x() == 1.0f);
    REQUIRE(cloud2.points()[1].y() == 5.0f);
    REQUIRE(cloud2.points()[2].z() == 9.0f);
}

TEST_CASE("PointCloud JSON multiple points") {
    std::vector<Point> points = {
        Point(1.0f, 2.0f, 3.0f),
        Point(4.0f, 5.0f, 6.0f),
        Point(7.0f, 8.0f, 9.0f)
    };
    std::vector<Vector> normals = {
        Vector(0.0f, 0.0f, 1.0f),
        Vector(0.0f, 1.0f, 0.0f),
        Vector(1.0f, 0.0f, 0.0f)
    };
    std::vector<Color> colors = {
        Color(255, 0, 0, 255),
        Color(0, 255, 0, 255),
        Color(0, 0, 255, 255)
    };
    PointCloud cloud(points, normals, colors);
    
    auto json = cloud.to_json_data();
    PointCloud cloud2 = PointCloud::from_json_data(json);
    
    REQUIRE(cloud2.size() == 3);
    REQUIRE(cloud2.points()[0].x() == 1.0f);
    REQUIRE(cloud2.points()[1].y() == 5.0f);
    REQUIRE(cloud2.points()[2].z() == 9.0f);
    REQUIRE(cloud2.normals()[0].z() == 1.0f);
    REQUIRE(cloud2.colors()[1].g == 255);
    // Verify alpha is always 255 after deserialization
    REQUIRE(cloud2.colors()[0].a == 255);
    REQUIRE(cloud2.colors()[1].a == 255);
    REQUIRE(cloud2.colors()[2].a == 255);
}
